hooli_data_eng/assets/dbt_assets.py

```diff
@@ -246,3 +253,56 @@ def dbt_slim_ci(dbt2: DbtCliResource):
 @job
 def dbt_slim_ci_job():
     dbt_slim_ci()
+
+
+# Load and parse the dbt manifest
+def load_dbt_manifest(manifest_path):
+    with open(manifest_path, 'r') as f:
+        manifest = json.load(f)
+    return manifest
+
+# Create AssetCheckSpec objects from the dbt manifest
+def create_source_freshness_check_specs(manifest_path):
+    manifest = load_dbt_manifest(manifest_path)
+    check_specs = []
+    for _, source_data in manifest['sources'].items():
+        if "freshness" in source_data and source_data["freshness"].get("error_after").get("count") is not None and source_data["name"] != "locations":
+            asset_key = AssetKey([source_data["schema"].upper(), source_data["name"] ])
+            check_specs.append(
+                AssetCheckSpec(
+                    name="source_freshness_check",
+                    asset=asset_key,
+                    blocking=True,
+                    automation_condition=allow_outdated_and_missing_parents_condition),
+            )
+    return check_specs
+
+@multi_asset_check(
+    specs=create_source_freshness_check_specs(dbt_project.manifest_path),
+    can_subset=True,
+    required_resource_keys={"dbt"},
+)
+def source_freshness_check(context: AssetCheckExecutionContext):
+    dbt_resource = context.resources.dbt
+
+    selection_list = [f"source:{asset_check_key.asset_key.path[0].lower()}.{asset_check_key.asset_key.path[1]}" for asset_check_key in context.selected_asset_check_keys]
+    selection_string = " ".join(selection_list)
+
+    context.log.info(f"Running check source_freshness_check on assets {selection_string}")
+    dbt_cli_invocation = dbt_resource.cli(["source", "freshness", "--select", selection_string], raise_on_error=False)
+    # TODO add better metadata
+    # Process the raw event stream
+    for dbt_event in dbt_cli_invocation.stream_raw_events():
+        context.log.info(f"Received event: {dbt_event.raw_event}")
+        if dbt_event.raw_event.get('info', {}).get('name') == 'LogFreshnessResult':
+            log_freshness_result = dbt_event.raw_event['info']
+            context.log.info(f"Filtered LogFreshnessResult: {log_freshness_result}")
+            passed = True if log_freshness_result['level'] == 'info' else False
+            severity = AssetCheckSeverity.ERROR if log_freshness_result['level'] == 'error' else AssetCheckSeverity.WARN
```

Comment:
This is very cool and something I think we should make more of a default practice across the integration now that blocking asset checks are more generally available with DA

Reply:
💯 ! They don't work with dbt tests on dbt assets I don't think? But will play around more with this as I'd like to understand it better

Reply:
IIUC dbt tests on dbt assets will pretty much always block other stuff in the same run, since they fail the dbt op... but idk about how they work with DA